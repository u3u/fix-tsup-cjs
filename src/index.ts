export * from './fix-cjs-dts'
export * from './fix-cjs-exports'
export * from './reflect'
