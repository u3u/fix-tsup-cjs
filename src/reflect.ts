import fs from 'node:fs/promises';
import path from 'node:path';
import fg from 'fast-glob';
import { gray, green } from 'kolorist';

export type ReflectActionResult = string | false | undefined;

export interface ReflectOptions {
  files: fg.Pattern | fg.Pattern[];
  globOptions?: fg.Options;
  name?: string;
  reflect(code: string): Promise<ReflectActionResult> | ReflectActionResult;
}

export const reflect = async (options: ReflectOptions) => {
  const { globOptions = {}, name = 'reflect', reflect } = options;
  const cwd = process.cwd();

  const files = await fg(options.files, {
    cwd: path.resolve(cwd, 'dist'),
    ...globOptions,
    absolute: true,
    ignore: [...(globOptions.ignore || []), 'node_modules'],
  });

  const logger = (message: string) => console.log(`[${name}]`, message);

  if (!files.length) logger(gray('No files matched'));

  const output = [];

  for (const file of files) {
    const code = await fs.readFile(file, 'utf8');
    const result = await reflect(code);
    const filename = path.relative(cwd, file);

    if (result) {
      await fs.writeFile(file, result);
      output.push(result);
      logger(`${green('âœ”')} ${filename}`);
    } else {
      logger(gray(`skip ${filename}`));
    }
  }

  return output;
};
