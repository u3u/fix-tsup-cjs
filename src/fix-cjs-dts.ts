import { reflect, type ReflectOptions } from './reflect';

const exportRegex = /export \{\s*(?:\S.*)?\b(?<name>.+) as default.*(?:[\n\r\u{2028}\u{2029}]\s*)?\};/u;

export const fixCjsDts = async (options?: Partial<ReflectOptions>) => {
  return reflect({
    files: '**/*.d.{ts,cts}',
    ...options,
    name: 'fix-cjs-dts',

    reflect: (code) => {
      const result = exportRegex.exec(code);

      if (result?.groups?.name) {
        const statement = `export = ${result.groups.name}`;

        if (!code.endsWith(statement)) return code + statement;
      }
    },
  });
};
