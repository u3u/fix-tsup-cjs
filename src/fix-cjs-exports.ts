import { reflect, type ReflectOptions } from './reflect'

const statement = `
// fix-cjs-exports
if (module.exports.default) {
  Object.assign(module.exports.default, module.exports);
  module.exports = module.exports.default;
  delete module.exports.default;
}
`

export const fixCjsExports = async (options?: Partial<ReflectOptions>) => {
  const { readPackage } = await import('read-pkg')
  const { type } = await readPackage()
  const isEsm = type === 'module'
  const suffix = isEsm ? 'cjs' : 'js'
  const files = options?.files?.length ? options.files : `**/*.${suffix}`

  await reflect({
    ...options,
    files,
    name: 'fix-cjs-exports',
    reflect: (code) => {
      if (code.includes('module.exports =') && !code.includes(statement)) {
        return code + statement
      }
    },
  })
}
