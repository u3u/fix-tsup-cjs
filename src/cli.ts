#!/usr/bin/env node
import path from 'node:path';
import { cac } from 'cac';
import { name, version } from '../package.json';
import { fixCjsDts } from './fix-cjs-dts';
import { fixCjsExports } from './fix-cjs-exports';
import { toArray } from './to-array';

const cli = cac(name);

cli.version(version);

cli
  .command('[...files]', 'Custom matching files glob')
  .option('--cwd [path]', 'Set fix directory', { default: 'dist' })
  .option('--dts', 'Fix commonjs d.ts files', { default: true })
  .option('-i, --ignore [...files]', 'Ignore files')
  .action(async (files, options) => {
    const { dts } = options;
    const hasCustomMatchFiles = !!files?.length;
    const currentDir = process.cwd();
    const cwd = path.resolve(currentDir, options.cwd);
    const ignore = toArray(options.ignore);

    await fixCjsExports({
      files,

      globOptions: {
        cwd: hasCustomMatchFiles ? currentDir : cwd,
        ignore,
      },
    });

    if (dts) {
      await fixCjsDts({
        globOptions: {
          cwd,
          ignore,
        },
      });
    }
  });

cli.help();

cli.parse();
