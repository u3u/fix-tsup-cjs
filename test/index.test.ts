import path from 'node:path';
import { expect, it } from 'vitest';
import { fixCjsDts, fixCjsExports } from '../src';

const entry = 'mod';
const outDir = 'temp';

const tsup = async () => {
  const { execa } = await import('execa');

  await execa('tsup', [
    //
    `test/__fixtures__/${entry}.ts`,
    '--clean',
    '--dts',
    '--format',
    'cjs,esm',
    '--out-dir',
    outDir,
    '--shims',
  ]);
};

it('fix-cjs-exports', async () => {
  await tsup();

  const result = await fixCjsExports({
    globOptions: {
      cwd: outDir,
    },
  });

  const mod = require(path.join('..', outDir, entry));

  expect(mod).toBeDefined();
  expect(mod).toMatchSnapshot();
  expect(result).toMatchSnapshot();
});

it('fix-cjs-dts', async () => {
  await tsup();

  const result = await fixCjsDts({
    globOptions: {
      cwd: outDir,
    },
  });

  expect(result).toMatchSnapshot();
});
