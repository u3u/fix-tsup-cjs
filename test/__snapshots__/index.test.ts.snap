// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fix-cjs-dts 1`] = `
[
  "declare const _default: {
    readonly format: \\"commonjs\\";
};

export { _default as default };
export = _default",
]
`;

exports[`fix-cjs-exports 1`] = `
{
  "format": "commonjs",
}
`;

exports[`fix-cjs-exports 2`] = `
[
  "\\"use strict\\";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === \\"object\\" || typeof from === \\"function\\") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, \\"__esModule\\", { value: true }), mod);

// test/__fixtures__/mod.ts
var mod_exports = {};
__export(mod_exports, {
  default: () => mod_default
});
module.exports = __toCommonJS(mod_exports);
var mod_default = {
  format: \\"commonjs\\"
};

// fix-cjs-exports
if (module.exports.default) {
  Object.assign(module.exports.default, module.exports);
  module.exports = module.exports.default;
  delete module.exports.default;
}
",
]
`;
